[INTEL MKL] Integrated MKL input conversion op with MKL-DNN v1.x.

bhavani-subramanian opened the pull request.

bhavani-subramanian committed:
Integrated MKL input conversion op with MKL-DNN v1.x.

penpornk commented on tensorflow/core/kernels/mkl_input_conversion_op.cc:
```diff
@@ -136,19 +143,32 @@ class MklInputConversionOp : public OpKernel {
+              context, status,
+              errors::Internal(
+                  "MklInputConversionOp: Failed to create reorder for input0"));
+          ExecutePrimitive(net, &net_args, cpu_engine);
```
Thank you for refactoring `ExecutePrimitive`! Let's `#define NET_ARGS_PTR` (to `&new_args` vs `nullptr`) so we can put `ExecutePrimitive` calls outside of `#ifdef`. `OP_REQUIRES` can be pulled off of `#ifdef` as well. 

Refactoring `CheckReorderToOpMem` sounds good. Thank you for adding the TODO! :)

bhavani-subramanian replied:
Thanks @penpornk! Defined another macro per your suggestion.

penpornk reviewed (changes requested):
Thank you for the PR! I have minor formatting comments.

bhavani-subramanian committed:
1. Replaced typedefs with using statements.
2. Added code to directly execute conditions within OP_REQUIRES.
3. Defined some more macros to help minimize code duplication.

bhavani-subramanian commented on tensorflow/core/kernels/mkl_conv_ops.cc:
```diff
@@ -53,6 +53,9 @@ using mkldnn::stream;
 
 namespace tensorflow {
 
+typedef mkldnn::convolution_forward::primitive_desc ConvFwdPd;
```
Done.

bhavani-subramanian commented on tensorflow/core/kernels/mkl_conv_ops.cc:
```diff
@@ -1007,35 +1010,46 @@ class MklConvOp : public OpKernel {
-        DCHECK(result);
+        auto status = (*output_tensor)->CopyFrom(add_tensor, output_tf_shape);
+        OP_REQUIRES(
+            context, status,
```
Done.

bhavani-subramanian commented on tensorflow/core/util/mkl_util.h:
```diff
@@ -667,30 +684,40 @@ inline Tensor ConvertMklToTF(OpKernelContext* context, const Tensor& mkl_tensor,
-               true);
-      stream(stream::kind::eager).submit(net).wait();
+      auto status =
+          input.CheckReorderToOpMem(output_tf_pd, &output_tensor, &net);
```
@penpornk When I tried using `OP_REQUIRES` here, I got a build error. I think `OP_REQUIRES` can only be called from functions that return `void` (https://github.com/tensorflow/tensorflow/blob/b5b258faa19d7ffc377a7322707e5098267198b3/tensorflow/core/framework/op_kernel.h#L1784).
Please let me know if I am missing something here.

bhavani-subramanian replied:
Sure, I'll do that! Thanks for the clarification.

bhavani-subramanian replied:
@penpornk I have changed `ConvertMklToTF()` to return the `Status`. Please take a look. Thanks.

bhavani-subramanian commented:
@penpornk I have addressed your review comments. Please take a look. Thanks!

penpornk reviewed (changes requested):
Thank you very much for the changes!

bhavani-subramanian committed:
Changed ConvertMklToTF() to return the status.

penpornk commented on tensorflow/core/util/mkl_util.h:
```diff
@@ -666,57 +666,63 @@ inline void ExecutePrimitive(const std::vector<primitive>& net,
+inline Status ConvertMklToTF(OpKernelContext* context,
+                             const Tensor& input_mkl_tensor,
+                             const MklDnnShape& input_mkl_shape,
+                             Tensor& output_tf_tensor) {
```
According to our [C++ style guide](https://google.github.io/styleguide/cppguide.html#Reference_Arguments), the output argument should be a pointer, i.e., `Tensor* output_tf_tensor` instead of `Tensor& output_tf_tensor`. Sorry for the inconvenience!

bhavani-subramanian replied:
@penpornk The one in `mkl_tfconv_op.h` is using `ConvertMklToTf()` and not `ConvertMklToTF()` (the former is defined locally). Same for `mkl_quantized_ops_test.cc` and `mkl_input_conversion_op.cc`.
I've updated the calls in `mkl_concat_op.cc` and `mkl_lrn_op.cc` since they are using the one I just modified.

bhavani-subramanian replied:
> According to our [C++ style guide](https://google.github.io/styleguide/cppguide.html#Reference_Arguments), the output argument should be a pointer, i.e., `Tensor* output_tf_tensor` instead of `Tensor& output_tf_tensor`. Sorry for the inconvenience!

No worries! I'll change it to be a pointer.

bhavani-subramanian replied:
@penpornk I have changed the output argument of `ConvertMklToTF()` to be a pointer. Please review. Thanks!

penpornk replied:
I think there is a misunderstanding. The output argument of `ConvertMklToTf()` should be a pointer and the function signature should be
```c++
inline Status ConvertMklToTF(OpKernelContext* context,
                             const Tensor& input_mkl_tensor,
                             const MklDnnShape& input_mkl_shape,
                             Tensor* output_tf_tensor) {
  ...
  *output_tf_tensor = ...;
}
```
instead of
```c++
inline Status ConvertMklToTF(OpKernelContext* context,
                             const Tensor& input_mkl_tensor,
                             const MklDnnShape& input_mkl_shape,
                             Tensor& output_tf_tensor) {
  ...
  output_tf_tensor = ...;
}
```


And `ConvertMklToTf` should be called like this
```c++
Tensor output;
ConvertMklToTf<T>(context, input_mkl_tensor, input_mkl_shape, &output);
```
instead of 
```c++
Tensor output;
ConvertMklToTf<T>(context, input_mkl_tensor, input_mkl_shape, output);
```

penpornk reviewed (changes requested):
Thank you very much for the changes! I did a quick search for `ConvertMklToTF` and it seems `mkl_tfconv_op.h`, `mkl_quantized_ops_test.cc`, `mkl_lrn_op.cc`, and `mkl_input_conversion_op.cc` are using it as well. Should you also change them? I'm just afraid this might break your build.

bhavani-subramanian committed:
Changed the output argument of ConvertMklToTF() to be a pointer.

bhavani-subramanian committed:
Adding missing changes to ConvertMklToTF()

bhavani-subramanian commented:
@penpornk Looks like some changes didn't get added when I copied them as a `diff` from my internal branch. I've added them now. Sorry for the inconvenience.

penpornk commented on tensorflow/core/kernels/mkl_concat_op.cc:
```diff
@@ -578,9 +575,11 @@ class MklConcatOp : public OpKernel {
+        Tensor* tmp_tensor;
+        OP_REQUIRES_OK(
+            context, ConvertMklToTF<T>(context, values[i], mkl_input_shapes[i],
+                                       tmp_tensor));
```
We can just call
```c++
OP_REQUIRES_OK(
    context, ConvertMklToTF<T>(context, values[i], mkl_input_shapes[i],
                               &converted_values[i]));
```
and eliminate `tmp_tensor`.

penpornk commented on tensorflow/core/kernels/mkl_lrn_op.cc:
```diff
@@ -108,9 +107,11 @@ class MklLRNOp : public OpKernel {
+        OP_REQUIRES_OK(context,
+                       ConvertMklToTF<T>(context, src_tensor, src_dnn_shape,
+                                         converted_tensor));
+        MklDefaultToEigen(context, *converted_tensor);
```
I think it's cleaner to make `converted_tensor` a `Tensor` and not `Tensor*`.
```c++
        Tensor converted_tensor;
        OP_REQUIRES_OK(context,
                       ConvertMklToTF<T>(context, src_tensor, src_dnn_shape,
                                         &converted_tensor));
        MklDefaultToEigen(context, converted_tensor);
```

bhavani-subramanian replied:
Sure, I'll make these changes.

penpornk reviewed (changes requested):
@bhavani-subramanian That makes sense. Thank you!
Do you have to change how `ConvertMklToTf` is called in `mkl_tfconv_op.h`, `mkl_quantized_ops_test.cc`, `mkl_lrn_op.cc`, and `mkl_input_conversion_op.cc` as well?

bhavani-subramanian commented:
@penpornk The one in `mkl_tfconv_op.h` takes additional parameters compared to the one in `mkl_util.h`. Do you want me to merge the two and have just one function in `mkl_util.h`?

Regarding `ConvertMklToTF()` in `mkl_quantized_conv_ops_test.cc`, do you want me to move that to `mkl_util.h` as well?

penpornk commented:
@bhavani-subramanian I see. So there are multiple `ConvertMklToTf()` in different files. We should unify them, or at least put all of them in the same place (`mkl_util.h`). I think we should have a separate PR for this. Could you please add a TODO? For this PR, I just want to make sure that the changes won't break the build.

bhavani-subramanian committed:
Addressed review comments.

bhavani-subramanian commented:
@penpornk I finished making the changes you requested. Thanks again for reviewing the PR :)

penpornk reviewed (approved):
Thank you for your patience as well! Have a great weekend! :)

mellanox-github commented:
Can one of the admins verify this patch?

bhavani-subramanian commented:
@penpornk Just checking if there's anything else needed from my end w.r.t this PR?

penpornk commented:
Sorry for the delay! This PR was pulled in and was being merged automatically. I didn't notice that merge was unsuccessful.

There is a merge conflict internally (with `mkl_util.h`). Could you please try syncing the PR and resolve the conflict? (If you can't see the conflict, let me know and I'll resolve this internally.)

bhavani-subramanian commented:
@penpornk No worries! I just tried to merge this branch with master and I don't see any merge conflicts.

penpornk commented:
@bhavani-subramanian Thanks for checking! I'll do it then. :)
